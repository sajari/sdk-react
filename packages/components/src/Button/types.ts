import { AriaButtonProps } from '@react-types/button';
import { PropsWithAs } from '@sajari/react-sdk-utils';

import { BoxProps } from '../Box';

export type ButtonAppearance = 'default' | 'primary' | 'link' | 'subtle-link' | 'none';
export type ButtonSize = 'xs' | 'sm' | 'md' | 'lg' | 'xl';
export type ButtonSpacing = 'compact' | 'default' | 'none';

interface Props extends BoxProps {
  /** The base styling to apply to the button */
  appearance?: ButtonAppearance;
  /** Set how much spacing the button should have */
  spacing?: ButtonSpacing;
  /** The size to apply to the button */
  size?: ButtonSize;
  /** Provides a url for buttons being used as a link */
  loading?: boolean;
  /** Set if the button should be full width */
  fullWidth?: boolean;
  /** The classname for pressed button */
  pressedClassName?: string;
}

export interface ButtonProps extends PropsWithAs<Props>, AriaButtonProps {}
