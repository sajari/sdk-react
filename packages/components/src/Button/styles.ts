import { css } from '@emotion/core';
import { useTheme } from 'sajari-react-sdk-styles';
import tw, { TwStyle } from 'twin.macro';

import { ButtonProps } from './types';

const useButtonStyles = ({
  size,
  fullWidth,
  appearance,
  spacing,
  disabled,
  loading,
  pressed,
  focused,
  hovered,
}: ButtonProps & { pressed: boolean; hovered: boolean; focused: boolean }) => {
  const theme = useTheme();
  const isLink = (['link', 'subtle-link'] as Array<ButtonProps['apperance']>).includes(appearance);
  const styles: (string | TwStyle)[] = [];

  styles.push(
    tw`items-center justify-center no-underline transition duration-150 ease-in-out border border-transparent border-solid cursor-pointer focus:outline-none`,
  );

  if (!isLink) {
    styles.push(tw`font-medium rounded-md select-none`);
  }

  styles.push(fullWidth ? tw`flex` : tw`inline-flex`);

  switch (size) {
    case 'xl':
      if (!isLink) {
        styles.push(tw`text-xl`);
      }
      if (spacing !== 'none') {
        if (spacing === 'compact') {
          styles.push(tw`px-8`);
        } else {
          styles.push(tw`px-12 py-4`);
        }
      }
      break;

    case 'lg':
      if (!isLink) {
        styles.push(tw`text-lg`);
      }
      if (spacing !== 'none') {
        if (spacing === 'compact') {
          styles.push(tw`px-6`);
        } else {
          styles.push(tw`px-8 py-3`);
        }
      }
      break;

    case 'sm':
      styles.push(tw`text-sm`);

      if (spacing !== 'none') {
        if (spacing === 'compact') {
          styles.push(tw`px-2`);
        } else {
          styles.push(tw`px-3 py-1`);
        }
      }
      break;

    case 'xs':
      styles.push(tw`text-xs`);

      if (spacing !== 'none') {
        if (spacing === 'compact') {
          styles.push(tw`px-2`);
        } else {
          styles.push(tw`px-3 py-1`);
        }
      }
      break;

    default:
    case 'md':
      if (!isLink) {
        styles.push(tw`text-base`);
      }
      if (spacing !== 'none') {
        if (spacing === 'compact') {
          styles.push(tw`px-3`);
        } else {
          styles.push(tw`px-4 py-2`);
        }
      }
      break;
  }

  if (disabled) {
    styles.push(tw`text-gray-400 cursor-not-allowed`);

    switch (appearance) {
      case 'default':
        styles.push(tw`bg-gray-100 border-gray-200`);
        break;

      default:
        break;
    }

    return css(styles);
  }

  switch (appearance) {
    // TODO: test theming for primary appearance, we might want to change them later
    case 'primary':
      styles.push(`color: ${theme.color.primary.text};`);

      if (focused || pressed) {
        styles.push(`background: ${theme.color.primary.active};`);
      } else if (hovered) {
        styles.push(`background: ${theme.color.primary.active};`);
      } else {
        styles.push(`background: ${theme.color.primary.base};`);
      }
      break;

    case 'default':
      styles.push(
        tw`text-gray-500 bg-white border-gray-200 shadow-sm focus:border-gray-400 hover:bg-gray-100 focus:bg-white focus:shadow-outline-gray`,
      );
      if (pressed) {
        styles.push(tw`bg-gray-100`);
      }
      break;

    case 'link':
      styles.push(
        tw`no-underline bg-transparent hover:underline focus:underline focus:shadow-outline focus:shadow-outline-blue`,
      );
      styles.push({ color: pressed ? theme.color.primary.active : theme.color.primary.base });

      if (!loading && !pressed) {
        styles.push(`&:hover, &:focus { color: ${theme.color.primary.active} }`);
      }
      break;

    case 'subtle-link':
      styles.push(
        tw`no-underline bg-transparent hover:underline focus:underline focus:shadow-outline focus:shadow-outline-gray`,
      );

      styles.push(pressed ? tw`text-gray-700` : tw`text-gray-500`);

      if (!loading && !pressed) {
        styles.push(tw`hover:text-gray-700 focus:text-gray-700`);
      }
      break;
    default:
      break;
  }

  return css(styles);
};

export default useButtonStyles;
