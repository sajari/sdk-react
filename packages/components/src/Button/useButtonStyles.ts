import { css } from '@emotion/core';
import tw, { TwStyle } from 'twin.macro';
import { useTheme } from 'sajari-react-sdk-styles';
import { ButtonProps } from './types';

const useButtonStyles = ({
  size,
  fullWidth,
  appearance,
  spacing,
  disabled,
  loading,
  pressed,
  focused,
  hovered,
}: ButtonProps & { pressed: boolean; hovered: boolean; focused: boolean }) => {
  const theme = useTheme();

  const isLink = (['link', 'subtle-link'] as Array<ButtonProps['apperance']>).includes(appearance);

  const styles: (string | TwStyle)[] = [];

  styles.push(
    tw`focus:outline-none items-center justify-center border-solid border transition duration-150 ease-in-out cursor-pointer no-underline`,
  );

  if (!isLink) {
    styles.push(tw`font-medium select-none rounded-md`);
  }

  styles.push(fullWidth ? tw`flex` : tw`inline-flex`);

  switch (size) {
    case 'xl':
      if (!isLink) {
        styles.push(tw`text-xl`);
      }
      if (spacing !== 'none') {
        if (spacing === 'compact') {
          styles.push(tw`px-8`);
        } else {
          styles.push(tw`px-12 py-4`);
        }
      }
      break;

    case 'lg':
      if (!isLink) {
        styles.push(tw`text-lg`);
      }
      if (spacing !== 'none') {
        if (spacing === 'compact') {
          styles.push(tw`px-6`);
        } else {
          styles.push(tw`px-8 py-3`);
        }
      }
      break;

    case 'sm':
      if (!isLink) {
        styles.push(tw`text-sm`);
      }
      if (spacing !== 'none') {
        if (spacing === 'compact') {
          styles.push(tw`px-2`);
        } else {
          styles.push(tw`px-3 py-1`);
        }
      }
      break;

    case 'xs':
      if (!isLink) {
        styles.push(tw`text-xs`);
      }
      if (spacing !== 'none') {
        if (spacing === 'compact') {
          styles.push(tw`px-2`);
        } else {
          styles.push(tw`px-3 py-1`);
        }
      }
      break;

    default:
    case 'md':
      if (!isLink) {
        styles.push(tw`text-base`);
      }
      if (spacing !== 'none') {
        if (spacing === 'compact') {
          styles.push(tw`px-3`);
        } else {
          styles.push(tw`px-4 py-2`);
        }
      }
      break;
  }

  if (disabled) {
    styles.push(tw`text-gray-400 cursor-not-allowed`);

    switch (appearance) {
      case 'default':
        styles.push(tw`border-gray-200 bg-gray-100`);
        break;

      default:
        styles.push(tw`border-transparent`);
        break;
    }

    return css(styles);
  }

  switch (appearance) {
    // TODO: test theming for primary appearance, we might want to change them later
    case 'primary':
      styles.push(`
        color: ${theme.color.primary.text};
        border-color: ${theme.color.primary.border};
      `);

      if (focused || pressed) {
        styles.push(`background: ${theme.color.primary.active};`);
        styles.push(`box-shadow: 0 0 0 3px ${theme.color.primary.outline};`);
      } else if (hovered) {
        styles.push(`background: ${theme.color.primary.active};`);
      } else {
        styles.push(`background: ${theme.color.primary.base};`);
      }
      break;

    case 'default':
      styles.push(
        tw`text-gray-500 border-gray-200 focus:border-gray-400 bg-white hover:bg-gray-100 focus:bg-white focus:border-opacity-75 focus:shadow-outline-gray shadow-sm`,
      );
      if (pressed) {
        styles.push(tw`bg-gray-100`);
      }
      break;

    case 'link':
      styles.push(
        tw`bg-transparent no-underline hover:underline focus:underline border-transparent focus:shadow-outline focus:shadow-outline-blue`,
      );
      styles.push(pressed ? tw`text-blue-700` : tw`text-blue-500`);

      if (!loading && !pressed) {
        styles.push(tw`hover:text-blue-700 focus:text-blue-700`);
      }
      break;

    case 'subtle-link':
      styles.push(
        tw`bg-transparent no-underline hover:underline focus:underline border-transparent focus:shadow-outline focus:shadow-outline-gray`,
      );

      styles.push(pressed ? tw`text-gray-700` : tw`text-gray-500`);

      if (!loading && !pressed) {
        styles.push(tw`hover:text-gray-700 focus:text-gray-700`);
      }
      break;
    default:
      break;
  }

  return css(styles);
};

export default useButtonStyles;
