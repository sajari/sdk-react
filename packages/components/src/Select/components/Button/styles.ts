import { mapStyles } from '@sajari/react-sdk-utils';
import tw from 'twin.macro';

import { UseInputStyleProps, useInputStyles } from '../../../hooks';
import { SelectProps } from '../../types';

export type UseButtonStylesParams = Required<Pick<SelectProps, 'size' | 'invalid' | 'disabled'>>;

export function useButtonStyles(params: UseButtonStylesParams) {
  const { styles: inputStyles, focusProps: focusRingProps, focusRingStyles } = useInputStyles({
    type: 'select',
    ...params,
  } as UseInputStyleProps);

  const styles = { container: [tw`form-select`, inputStyles] };

  return { styles: mapStyles(styles), focusRingProps, focusRingStyles };
}
