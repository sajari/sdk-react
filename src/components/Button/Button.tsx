import React, { HTMLAttributes } from "react";
import tw, { styled } from "twin.macro";

const StyledButton = styled.button`
  ${tw`px-5 py-2 border-none text-white`};
  background-color: ${({ theme }) => theme.color.primary};
`;

export const Button = (props: HTMLAttributes<HTMLButtonElement>) => {
  return <StyledButton {...props} />;
};
