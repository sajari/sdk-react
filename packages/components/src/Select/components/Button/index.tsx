import { mergeProps } from '@react-aria/utils';
import { getStylesObject } from '@sajari/react-sdk-utils';
import * as React from 'react';

import Box from '../../../Box';
import { useSelectContext } from '../../context';
import { useButtonStyles } from './styles';

const Button = () => {
  const {
    disabled,
    getToggleButtonProps,
    id,
    invalid,
    items,
    selectedItems,
    autofocus,
    size,
    text,
    disableDefaultStyles,
    customClassNames,
  } = useSelectContext();
  const { styles: buttonStyles, focusRingProps, focusRingStyles } = useButtonStyles({ size, disabled, invalid });
  const styles = getStylesObject(buttonStyles, disableDefaultStyles);
  const ref = React.useRef<HTMLButtonElement>(null);

  React.useEffect(() => {
    if (autofocus && ref?.current) {
      ref.current.focus();
    }
  }, []);

  const children: React.ReactNode = React.useMemo(() => {
    if (typeof text === 'function') {
      const selected = items.filter(([v]) => selectedItems.includes(v.toString())).map(([, label]) => label);
      return text(selected);
    }

    if (typeof text !== 'undefined') {
      return text;
    }

    if (selectedItems.length === 1) {
      const [selectedItem] = selectedItems;
      const item = items.find(([v]) => selectedItem === v.toString());

      if (item) {
        const [, t] = item;
        return t;
      }
    }

    if (selectedItems.length > 1) {
      return `${selectedItems.length} selected`;
    }

    return 'Select an option';
  }, [JSON.stringify(selectedItems), text]);

  return (
    <Box
      as="button"
      type="button"
      css={[styles.container, focusRingStyles]}
      className={customClassNames.buttonClassName}
      {...mergeProps(focusRingProps, getToggleButtonProps({ disabled, ref }), { id })}
    >
      {children}
    </Box>
  );
};

export default Button;
