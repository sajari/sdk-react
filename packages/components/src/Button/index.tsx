/** @jsx jsx */
import { jsx } from '@emotion/core';
import { useButton } from '@react-aria/button';
import { useFocus, useHover } from '@react-aria/interactions';
import { mergeProps } from '@react-aria/utils';
import { getStylesObject, mergeRefs } from '@sajari/react-sdk-utils';
import classnames from 'classnames';
import React, { useRef } from 'react';

import Box from '../Box';
import useButtonStyles from './styles';
import { ButtonProps } from './types';

const Button = React.forwardRef((props: ButtonProps, ref?: React.Ref<HTMLElement>) => {
  const {
    // TODO: handle the state later as we might need a Spinner/Icon component
    loading = false,
    as = 'button',
    appearance = 'default',
    size = 'md',
    spacing = 'default',
    disabled,
    fullWidth = false,
    onClick,
    children,
    className,
    styles: stylesProps,
    pressedClassName = '',
    disableDefaultStyles = false,
    ...rest
  } = props;

  const [focused, setFocused] = React.useState(false);
  const buttonRef = useRef<HTMLElement | null>(null);
  const ownRef = mergeRefs(buttonRef, ref);

  const { buttonProps, isPressed: pressed } = useButton(
    { ...rest, isDisabled: disabled, elementType: as, onPress: onClick },
    buttonRef,
  );
  const { hoverProps, isHovered: hovered } = useHover({
    isDisabled: disabled,
  });
  const { focusProps } = useFocus({
    isDisabled: disabled,
    onFocus: () => setFocused(true),
    onBlur: () => setFocused(false),
  });

  const { styles: containerStyles, focusRingProps } = useButtonStyles({
    pressed,
    appearance,
    disabled,
    fullWidth,
    loading,
    size,
    spacing,
    focused,
    hovered,
  });

  const customProps = mergeProps(buttonProps, focusProps, hoverProps, focusRingProps);
  const styles = getStylesObject({ container: containerStyles }, disableDefaultStyles);

  return (
    <Box
      as={as}
      ref={ownRef}
      {...customProps}
      css={[styles.container, stylesProps]}
      className={classnames(className, {
        [pressedClassName]: pressed,
      })}
    >
      {children}
    </Box>
  );
});

export default Button;
export type { ButtonProps };
