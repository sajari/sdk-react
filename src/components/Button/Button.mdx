---
name: Button
menu: Components
route: /components/button
---

import { ExampleContainer, categoryMultiFilter } from "../../docs/helpers.ts";
import ThemeProvider from "../../styles/theming";
import Button from ".";

# Button

## Usage

<ExampleContainer>
  <ThemeProvider>
    <Button>Default</Button>
    <Button appearance="primary">Primary</Button>
    <Button
      appearance="link"
      as="a"
      href="https://sajari.com"
      target="_blank"
    >
      Link
    </Button>
    <Button
      disabled
      appearance="link"
      as="a"
      href="https://sajari.com"
      target="_blank"
    >
      Disabled link
    </Button>
    <Button appearance="subtle-link" onFocus={() => console.log('focus')}>Subtle link</Button>
  </ThemeProvider>
</ExampleContainer>
