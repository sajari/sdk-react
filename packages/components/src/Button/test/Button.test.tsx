import { ThemeProvider } from '@sajari/react-sdk-utils';
import { fireEvent, render } from '@testing-library/react';
import * as React from 'react';

import Button from '..';

describe('Button', () => {
  test('Should match snapshot', () => {
    const { asFragment } = render(
      <ThemeProvider>
        <Button>test</Button>
      </ThemeProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should trigger click event', () => {
    const onClickSpy = jest.fn();
    const { getByRole } = render(
      <ThemeProvider>
        <Button onClick={onClickSpy}>Click me</Button>
      </ThemeProvider>,
    );
    const button = getByRole('button');
    fireEvent.click(button);

    expect(onClickSpy).toHaveBeenCalledTimes(1);
  });

  it('Should not response if disabled', () => {
    const onClickSpy = jest.fn();
    const { getByRole } = render(
      <ThemeProvider>
        <Button onClick={onClickSpy} disabled>
          Click me
        </Button>
      </ThemeProvider>,
    );
    const button = getByRole('button');
    fireEvent.click(button);

    expect(onClickSpy).not.toHaveBeenCalled();
  });

  it('Can have elementType=a with an href', () => {
    const onClickSpy = jest.fn();
    const { getByRole } = render(
      <ThemeProvider>
        <Button onClick={onClickSpy} href="sajari.com" as="a">
          Click me
        </Button>
      </ThemeProvider>,
    );

    const button = getByRole('button');
    expect(button).toHaveAttribute('tabindex', '0');
    expect(button).toHaveAttribute('href', 'sajari.com');
    fireEvent.click(button);
    expect(onClickSpy).toHaveBeenCalledTimes(1);
  });
});
