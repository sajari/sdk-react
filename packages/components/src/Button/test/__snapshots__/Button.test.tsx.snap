// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Should match snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1.5;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,-webkit-transform;
  -webkit-transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  border-width: 1px;
  border-color: transparent;
  border-style: solid;
  cursor: pointer;
  font-family: inherit;
  font-weight: 500;
  border-radius: 0.375rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  --text-opacity: 1;
  color: rgba(107,114,128,var(--text-opacity));
  --bg-opacity: 1;
  background-color: rgba(255,255,255,var(--bg-opacity));
  --border-opacity: 1;
  border-color: rgba(229,231,235,var(--border-opacity));
  box-shadow: 0 1px 2px 0 rgba(0,0,0,0.05);
}

.emotion-0:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
}

.emotion-0:hover {
  --bg-opacity: 1;
  background-color: rgba(244,245,247,var(--bg-opacity));
}

<button
    class=" emotion-0"
    type="button"
  >
    test
  </button>
</DocumentFragment>
`;
